docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+
+For other controllers in the controlplane (e.g MCM, etcd backup restore, kube apiserver) leader election has to be implemented. We plan to introduce gardenlet soon therefore the garden cluster is expected to be reachable from all seeds. Initially the Shoot's `spec.seedName` can be used to determine which seed is the current leader of the shoot.
+Another flag might be needed to tell the gardenlet in the _New Seed_ that the control plane in the _Host Seed_ has been scaled down and the Shoot reconciliation can begin.
+
+Horizontal Pod Autoscalers also need to be considered and removed.
+
+When discussing this section consider what happens in kubernetes:
+1. If the kubelet on a node can reach the apiserver it gets information about which pods are deployed on it and which should be evicted and moved to another node
+2. If the kubelet cannot connect to the apiserver, the kube controller manager will recreate pods on a new available node. However, containers already running on the disconnected node will not be shut down.
+3. If there is a PVC attached to a pod initially running on the broken node, the kube controller manager will recreate the pod on a new node and after some time (maxWaitForUnmountDuration) detach the volume forcefully so that it can be attached to the new pod.
+
+Possible ideas:
+- If both the Garden cluster and Host Seed cluster are healthy, the Garden Controller Manager can directly scale down the Shoot's control plane as part of the migration flow.
+- In other cases:
+    1. The `spec.seedName` in the Shoot resource can serve as leader election.
+        - The `gardenlet` (once it is implemented) can read it and scale down the control plane.
+        - Various mutationgwebhookconfigurations can create sidecar containers in the control plane controllers' pods. The sidecar containers will check whether the current Seed is a leader and if it is not, will shut down the networking on the pod.
+
+    2. Use a timestamp of succesfull reconciliation for Leader Election
+        - Once a Shoot is successfully reconciled on the Host Seed assign the Shoot to it by adding an assignment timestamp in the ShootState.
+        - The timestamp is propagated to the Cluster resource and to a sidecar running in the ETCD pod.
+        - The ETCD pod sidecar checks the backup bucket if there already is a timestap
+        - If there isn't it writes the new timestamp in the ETCD
+        - If there is an older timestamp the sidecar updates the cluster resource with an annotaiton specifying that the current Seed (where etcd is running) must start handling the Shoot's control plane
+        - If there is a newer timestamp the sidecar updates the cluster resource with an annotation specifying that the current Seed must stop handling the Shoot's control plane
+        - If the timestamps are the same nothing happens
+        - Other controllers can either get information from gardener cluster or the cluster resource
```

Comment:
As said, all of this has the problem that a healthy cluster shall not fall apart too easily. A single endpoint, the garden cluster is not reliable enough and we could reach a melt-down situation quickly if seeds can't reach the Gardener API server anymore and all breaks down (e.g. just because the LB is down or the DNS entry was lost - maybe due to human error). The outcome would be catastrophic.